### PR TITLE
feat: callbacks

### PR DESCRIPTION
**Pull request by 0xiamflux**
## Callbacks PoC

This draft showcases a complete callback flow where a call originates on the origin chain, triggers execution on a destination chain, and then returns the result to the origin chain, where it is handled with custom logic.

The flow is as follows:
image

### Possible Next Steps

* Having a dedicated library to abstract functionality and facilitate the interaction would be a valuable addition to the ecosystem.

### Possible Future Improvements
* Flexible Callback Contexts: Introducing a more complex data structure to support features like expiration times or additional metadata.
* Robust Error Handling: Supporting logic to handle both successful and failed callbacks.

### CODE REVIEW

File: packages/contracts-bedrock/src/L2/Callbacks.sol [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        bytes calldata _sentMessage
+    )
+        external
+        payable
+        returns (bytes memory returnData_)
+    {
+        // Calls the CDM contract and get return value of the function call
+        returnData_ = IL2ToL2CrossDomainMessenger(L2_TO_L2_CROSS_DOMAIN_MESSENGER).relayMessage(_id, _sentMessage);
+
+        (address sender,,) = abi.decode(_sentMessage[128:], (address, bytes, address));
```
Comment (skeletor-spaceman) [inline packages/contracts-bedrock/src/L2/Callbacks.sol]: why 128? could we add a comment on what this line does?

**Review by skeletor-spaceman — CHANGES_REQUESTED** [review]

File: packages/contracts-bedrock/src/L2/Callbacks.sol (hunk 1 above)
Comment (skeletor-spaceman) [reply packages/contracts-bedrock/src/L2/Callbacks.sol]: Can we just cut the 20bytes of the address?
128:148
Or does that not work?

Comment (0xiamflux) [reply packages/contracts-bedrock/src/L2/Callbacks.sol]: Yeah, changed to only decode the sender address, has to be 128:160, tho.

**Review by skeletor-spaceman — DISMISSED** [review]

**Review by skeletor-spaceman — DISMISSED** [review]

File: packages/contracts-bedrock/scripts/checks/interfaces/main.go [hunk 2]
```diff
@@ -318,7 +318,13 @@ func abiItemLess(a, b map[string]interface{}) bool {
...
 	aName := getString(a, "name")
 	bName := getString(b, "name")
-	return aName < bName
+	if aName != bName {
+		return aName < bName
+	}
+
+	aInputs := a["inputs"].([]interface{})
+	bInputs := b["inputs"].([]interface{})
+	return len(aInputs) < len(bInputs)
```
Comment (skeletor-spaceman) [inline packages/contracts-bedrock/scripts/checks/interfaces/main.go]: this is a weird change

File: packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol [hunk 3]
```diff
@@ -12,6 +12,8 @@ import { ISemver } from "interfaces/universal/ISemver.sol";
 import { IDependencySet } from "interfaces/L2/IDependencySet.sol";
 import { ICrossL2Inbox, Identifier } from "interfaces/L2/ICrossL2Inbox.sol";
 
+import "forge-std/Test.sol";
+
```
Comment (skeletor-spaceman) [inline packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol]: needed?

**Comment by 0xDiscotech** [comment]
Only focus on `Callbacks` and `Callbacks.t.sol` files. The other diff won't be a diff anymore once the entrypoints branch is merged on the OP repo.
Will merge with `develop` by then to solve any possible conflict or any other difference

File: packages/contracts-bedrock/src/L2/Callbacks.sol [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+}
+
+/**
+ * @title CallbackGreeter
+ * @notice Initiates a remote greeting call to a greeter contract on a different chain via the CallbackEntrypoint. It
+ * also handles the callback received from the remote greeter.
+ */
+contract CallbackGreeter is ICallbackGreeterMetadata {
+    /// @notice Address of the L2ToL2CrossDomainMessenger contract
+    address public constant L2_TO_L2_CROSS_DOMAIN_MESSENGER = Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER;
```
Comment (hexshire) [inline packages/contracts-bedrock/src/L2/Callbacks.sol:46]: Looks like there is not much consistency in the usage of predeploys addresses. I would like to use the SuperTokenBridge one, where this constant is called simply `MESSENGER`

https://github.com/ethereum-optimism/optimism/blob/0d0f69d59cfbce954710636efdf9812d9a38fead/packages/contracts-bedrock/src/L2/SuperchainTokenBridge.sol#L46

https://github.com/ethereum-optimism/optimism/blob/0d0f69d59cfbce954710636efdf9812d9a38fead/packages/contracts-bedrock/src/L2/SuperchainWETH.sol#L151

**Review by skeletor-spaceman — DISMISSED** [review]

**Review by hexshire — DISMISSED** [review]

**Comment by 0xDiscotech** [comment]
Closed in favor of a cleaner PR https://github.com/defi-wonderland/optimism/pull/162

**Pull request closed without merging**